src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -106,60 +143,250 @@ public IList Get(QueryKey key, ICacheAssembler[] returnTypes, bool isNaturalKeyL
 			{
 				if (returnTypes.Length == 1)
 				{
-					returnTypes[0]?.BeforeAssemble(cacheable[i], session);
+					cacheable.Add(returnTypes[0].Disassemble(row, session, null));
 				}
 				else
 				{
-					TypeHelper.BeforeAssemble((object[])cacheable[i], returnTypes, session);
+					cacheable.Add(TypeHelper.Disassemble((object[])row, returnTypes, null, session, null));
 				}
 			}
 
-			IList result = new List<object>(cacheable.Count - 1);
-			for (int i = 1; i < cacheable.Count; i++)
+			return cacheable;
+		}
+
+		private IList GetResultFromCacheable(
+			QueryKey key,
+			ICacheAssembler[] returnTypes,
+			bool isNaturalKeyLookup,
+			ISessionImplementor session,
+			IList cacheable)
+		{
+			Log.Debug("returning cached query results for: {0}", key);
+			if (key.ResultTransformer?.AutoDiscoverTypes == true && cacheable.Count > 0)
 			{
-				try
+				returnTypes = GuessTypes(cacheable);
+			}
+
+			try
+			{
+				var result = new List<object>(cacheable.Count - 1);
+				if (returnTypes.Length == 1)
 				{
-					if (returnTypes.Length == 1)
+					var returnType = returnTypes[0];
+
+					// Skip first element, it is the timestamp
+					var rows = new List<object>(cacheable.Count - 1);
+					for (var i = 1; i < cacheable.Count; i++)
 					{
-						result.Add(returnTypes[0].Assemble(cacheable[i], session, null));
+						rows.Add(cacheable[i]);
 					}
-					else
+
+					foreach (var row in rows)
 					{
-						result.Add(TypeHelper.Assemble((object[])cacheable[i], returnTypes, session, null));
+						returnType.BeforeAssemble(row, session);
+					}
+
+					foreach (var row in rows)
+					{
+						result.Add(returnType.Assemble(row, session, null));
 					}
 				}
-				catch (UnresolvableObjectException ex)
+				else
 				{
-					if (isNaturalKeyLookup)
+					// Skip first element, it is the timestamp
+					var rows = new List<object[]>(cacheable.Count - 1);
+					for (var i = 1; i < cacheable.Count; i++)
 					{
-						//TODO: not really completely correct, since
-						//      the UnresolvableObjectException could occur while resolving
-						//      associations, leaving the PC in an inconsistent state
-						Log.Debug(ex, "could not reassemble cached result set");
-						_queryCache.Remove(key);
-						return null;
+						rows.Add((object[]) cacheable[i]);
 					}
 
-					throw;
+					foreach (var row in rows)
+					{
+						TypeHelper.BeforeAssemble(row, returnTypes, session);
+					}
+
+					foreach (var row in rows)
+					{
+						result.Add(TypeHelper.Assemble(row, returnTypes, session, null));
+					}
 				}
+
+				return result;
 			}
+			catch (UnresolvableObjectException ex)
+			{
+				if (isNaturalKeyLookup)
+				{
+					//TODO: not really completely correct, since
+					//      the UnresolvableObjectException could occur while resolving
+					//      associations, leaving the PC in an inconsistent state
+					Log.Debug(ex, "could not reassemble cached result set");
+					// TODO: add and handle a RemoveMany?
```

Comment:
I forgot about `RemoveMany` when doing #1633 as it was not needed there, but it is [already implemented](https://github.com/nhibernate/NHibernate-Caches/pull/45/files#diff-bd639a7bcffa3fc62ab405a2ab4a63d5R353) in nhibernate/NHibernate-Caches#45, so why not add it if it can be used here.

Reply:
I have put a question mark because this case is almost an error case, and it does not look likely to me it would benefit from batching. It should not happen frequently, and even less frequently would it benefit from a `Many` handling, since the first encountered one short-circuit the result-set processing: it could be batched only in multi-queries case, if many queries in it encounter a "no more resolvable entity reference" case.

So in fact I should change that comment, I do not think it would be worth it to handle a many case here.